Closes #1683: Send fretboard experiment branch in telemetry

dnarcese opened the pull request:
### Pull Request checklist
<!-- Before submitting the PR, please address each item -->
- [x] This PR includes thorough **tests** or an explanation of why it does not
Telemetry change
- [x] This PR includes a **CHANGELOG entry** or does not need one

dnarcese commented:
# Request for data collection review form

1) What questions will you answer with this data?

- Which variant of the A/B test is the user in?

2) Why does Mozilla need to answer these questions?  Are there benefits for users? Do we need this information to address product or business requirements?

* To see and compare metrics from each variant of the split test to determine which is better for the users.

3) What alternative methods did you consider to answer these questions? Why were they not sufficient?

- This is the only way to determine this information.

4) Can current instrumentation answer these questions?

- No telemetry currently exists to answer these questions.

5) List all proposed measurements and indicate the category of data collection for each measurement, using the Firefox [data collection categories](https://wiki.mozilla.org/Firefox/Data_Collection) on the Mozilla wiki.   

| Measurement Description                      | Data Collection Category | Tracking Bug #                                               |
| -------------------------------------------- | ------------------------ | ------------------------------------------------------------ |
| Which variant of the A/B test the user is in | 2                        | [Issue 1683](https://github.com/mozilla-mobile/firefox-tv/issues/1683) |

6) How long will this data be collected?

* @athomasmoz will permanently monitor this data.

7) What populations will you measure?

* All users with telemetry enabled, on release channel.


8) If this data collection is default on, what is the opt-out mechanism for users?

- There is an option in settings to stop sending usage data.

9) Please provide a general description of how you will analyze this data.

- The data will be analyzed through Redash queries and dashboards.

10) Where do you intend to share the results of your analysis?

- On Redash and with mobile teams.

liuche commented on app/src/main/java/org/mozilla/tv/firefox/experiments/FretboardProvider.kt:
```diff
@@ -31,13 +31,25 @@ class FretboardProvider(private val applicationContext: Context) : CoroutineScop
+    // If the user is in the experiment (active), they are in branch B, otherwise branch A.
+    private val experimentBranches: List<String>
+        get() = fretboard.getExperimentsMap(applicationContext)
+                .map { it.key + if (it.value) ":B" else ":A" }
```
Also for future thought - I just talked to @suyounghong about how the Data team handles naming conventions for experiments, and he said in Normandy (Desktop experiments) usually there's a "tag" for the experiment, a bug number to uniquely cross-ref, and a "branch", so the whole experiment name might look like "onboarding_bug123456_b", which seems like something we should emulate for whatever our backend does.

SuYoungHong replied:
to elaborate on this. we handle the experiment and the branch as separate fields in telemetry. so, for example, if we have 2 branches in experiment-XXXX (where XXXX is the number of the bugzilla bug that launched the experiment), then one profile would have: 

```
{"experiment_id": "experiment-XXXX", "branch": "control"}
```
whereas another profile would have: 
```
{"experiment_id": "experiment-XXXX", "branch": "variant1"}
```

if it is the case that we can only use one field, and needed to combine the experiment_id and the branch, I would recommend using a unique deliminator like: 

```
{"experiment_id_and_branch": "experiment-XXXX:control"}
```

mmccorks replied:
Just the active experiments is fine. 

I'd like to keep the naming convention as close to desktop as possible.  Currently focus has something like `{"experiment-name":"branch"}` but having a bug reference in the name as well is useful. It might be a good idea to check in with @colintheshots as to how he implemented this in Focus.

dnarcese replied:
I interpreted 'active' to mean whether you are in the experiment. So I thought that `getExperimentsMap` gets both branches - those in the experiment, B, and those not, A. But `getActiveExperiments` would only get the users that are in branch B.

How do you run an experiment with more than two options? My understanding was that you specify one range of buckets, so you are either in the experiment or you are not.

I could be understanding things completely wrong - @colintheshots would you mind verifying/clarifying?

psymoon replied:
@liuche I think 'active' experiment means the user is in that experiment as per `fretboard#getActiveExperiments` implementation: 

`fun getActiveExperiments(context: Context): List<Experiment> { return experiments.filter { isInExperiment(context, ExperimentDescriptor(it.name)) }}`

liuche replied:
Right, I think that makes sense that we'd want to collect "active experiments". Since we're also creating the experiments ourselves, we know what *other* experiments this user isn't in. With every release, a user is being shown some version of an experiment, even if that is the control. So we'd still be able to measure usage by comparing users who are in "onboarding-a" and "onboarding-b" where either could be the control.

I guess I don't understand why we'd need to collect the experiments that a users **isn't** in?

dnarcese replied:
Doesn't not being in the experiment mean they are in branch A? So if we only get active experiments, that would always be branch B, so then we would never tag the control branch.

mmccorks replied:
There's a difference between active experiments and experiment branches. A user can be actively in an experiment but on the control branch (no treatment).  We need a list of the active experiments that a user is in as well as which branch (A, B, C...) they are in for each active experiment.

mmccorks replied:
Not all users will necessarily be in every experiment. Most experiments will only be active for some subset of the total population. Within that subset, the users will be assigned to a particular branch. So we need to have both the experiment id and the branch name. Those not in the experiment will not be tagged with the active experiment or a branch.

liuche replied:
Interesting! Okay, my only question that @dnarcese made me think of with this approach is, if the users not in an experiment don't have any tag (e.g., control for experimentX), how do you draw conclusions comparing the two populations (e.g. experiment outperformed control) when the control group for an experiment doesn't have a tag or something distinguishing them from users who, say, are stuck on an older version and also don't have any experiment tags? Can you filter the data based on app release version (e.g., v4.2+ only because that's when we release experiment X) or other ways of distinguishing?
Thanks for all the context @mmccorks! If just collecting activeExperiments is enough, let's do that - I just wanted to double check to make sure my understanding was correct.

mmccorks replied:
The two populations we compare are the two branches - the control group will be tagged as one of the branches. We need them explicitly tagged as part of the experiment to be able to draw conclusions like you said - with the correct version etc. so that we are only comparing people explicitly in the experiment. People in the control are IN the experiment, even if they are not receiving a treatment.

mmccorks replied:
We need the active experiments, and within those active experiments, which branch the user is on in that experiment.

dnarcese commented:
So I've changed implementation to only record active experiments. I've also updated the experiment config to include the branch in the name of the experiment. So the format is `experimentName-XXXX:branch` where XXXX will be the related GitHub issue.

liuche commented:
1) Is there or will there be **documentation** that describes the schema for the ultimate data set available publicly, complete and accurate?

There is documentation about the experiments format: https://github.com/mozilla-mobile/android-components/blob/master/components/service/fretboard/README.md#experiments-format-for-kinto

However, @dnarcese can you add a section on experiments in docs/telemetry.md, giving a brief explanation of the parts that we're using (Kinto backend, A/B switching client-side via Fretboard), and provide the format of the experiment config that you described above:
`experimentName-XXXX:branch where XXXX will be the related GitHub issue and the branch will be a letter where A is always control.`

2) Is there a control mechanism that allows the user to turn the data collection on and off? (Note, for data collection not needed for security purposes, Mozilla provides such a control mechanism) Provide details as to the control mechanism available.
Yes, users can turn off data with the data controls in Settings

3) If the request is for permanent data collection, is there someone who will monitor the data over time?
@athomasmoz will monitor the experiments data as they are run.

4) Using the **[category system of data types](https://wiki.mozilla.org/Firefox/Data_Collection)** on the Mozilla wiki, what collection type of data do the requested measurements fall under?  **
Type 2 - the type of experiment a user is shown

5) Is the data collection request for default-on or default-off?
Default on

6) Does the instrumentation include the addition of **any *new* identifiers** (whether anonymous or otherwise; e.g., username, random IDs, etc.  See the appendix for more details)?
No, this reuses the telemetry clientID like Focus, so that the Metrics team can verify bucketing is working, and sends up whether a user is in an existing bucket.
We've also discussed using a separate randomized id that is not tied to the telemetry clientID with @georgf in Glean for experiments only, but that is not present in this issue.

7) Is the data collection covered by the existing Firefox privacy notice? **If unsure: escalate to legal if:**
Yes, collecting the A/B testing bucket that a user is in, as determined by our A/B testing server.

8) Does there need to be a check-in in the future to determine whether to renew the data? (Yes/No) (If yes, set a todo reminder or file a bug if appropriate)**
No, experiments are part of the telemetry base ping.

dnarcese commented:
> Looks good! Do there need to be any changes for the FretboardProvider? It looks like that PR was merged in, so if there are changes, could be helpful to file to track the work.

I talked to @psymoon about the needed changes - I believe he's working on it. @psymoon can you file a bug to track it?
